File: 52. N-Queens II/52. N-Queens II.rs
impl Solution {
    pub fn total_n_queens(n: i32) -> i32 {
        let mut board = vec![0; n as usize];
        Self::solve_n_queens(&mut board, 0, n as usize) as i32
    }
    
    fn solve_n_queens(board: &mut Vec<i32>, row: usize, n: usize) -> usize {
        if row == n {
            return 1;
        }
        
        let mut count = 0;
        for col in 0..n {
            if Self::is_safe(board, row, col) {
                board[row] = col as i32;
                count += Self::solve_n_queens(board, row + 1, n);
            }
        }
        
        count
    }
    
    fn is_safe(board: &Vec<i32>, row: usize, col: usize) -> bool {
        for i in 0..row {
            if board[i] == col as i32 || 
               (board[i] - i as i32) == (col as i32 - row as i32) || 
               (board[i] + i as i32) == (col as i32 + row as i32) {
                return false;
            }
        }
        true
    }
}

File: 52. N-Queens II/52. N-Queens II.java
class Solution {
    public int totalNQueens(int n) {
        int[] board = new int[n];
        return solveNQueens(board, 0, n);
    }
    
    private int solveNQueens(int[] board, int row, int n) {
        if (row == n) {
            return 1;
        }
        
        int count = 0;
        for (int col = 0; col < n; col++) {
            if (isSafe(board, row, col)) {
                board[row] = col;
                count += solveNQueens(board, row + 1, n);
            }
        }
        
        return count;
    }
    
    private boolean isSafe(int[] board, int row, int col) {
        for (int i = 0; i < row; i++) {
            if (board[i] == col || 
                board[i] - i == col - row || 
                board[i] + i == col + row) {
                return false;
            }
        }
        return true;
    }
}